### PR TITLE
Feature/risk on develop

### PR DESCRIPTION
**Pull request by jerrylib**

### CODE REVIEW

**Review by mingshengke — APPROVED** [review]

File: contracts/riskon/UniswapV3UsdcWeth500RiskOnVault.sol [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import './UniswapV3RiskOnVault.sol';
+import './IUniswapV3RiskOnVaultInitialize.sol';
+
+contract UniswapV3UsdcWeth500RiskOnVault is IUniswapV3RiskOnVaultInitialize, UniswapV3RiskOnVault {
+
+    function initialize(address _owner, address _wantToken, address _uniswapV3RiskOnHelper, address _treasury, address _accessControlProxy) public override initializer {
+        super._initialize(
+            _owner,
+            _wantToken,
+        // https://info.uniswap.org/#/polygon/pools/0x45dda9cb7c25131df268515131f647d726f50608
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3UsdcWeth500RiskOnVault.sol:14]: add comment "WETH-USDC-UniswapV3Pool on polygon"

File: contracts/riskon/UniswapV3RiskOnHelper.sol [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    /// @notice Gets the specifie ranges of `_tick`
+    /// @param _tick The input number of tick
+    /// @param _tickSpacing The tick spacing
+    /// @param _baseThreshold The base threshold
+    /// @return _tickFloor The nearest tick which LTE `_tick`
+    /// @return _tickCeil The nearest tick which GTE `_tick`
+    /// @return _tickLower  `_tickFloor` subtrace `baseThreshold`
+    /// @return _tickUpper  `_tickFloor` add `baseThreshold`
+    function getSpecifiedRangesOfTick(int24 _tick, int24 _tickSpacing, int24 _baseThreshold) public view returns (int24 _tickFloor, int24 _tickCeil, int24 _tickLower, int24 _tickUpper) {
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnHelper.sol]: replace 'view' with 'pure'

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+            _claimAmounts[0] += _amount0;
+            _claimAmounts[1] += _amount1;
+        }
+
+        if (profitFeeBps > 0 && address(treasury) != address(0)) {
+            if (_claimAmounts[0] > 0) {
+                uint256 claimAmount0Fee = _claimAmounts[0] * profitFeeBps / 10000;
+                _claimAmounts[0] -= claimAmount0Fee;
+                IERC20Upgradeable(token0()).safeApprove(address(treasury), 0);
+                IERC20Upgradeable(token0()).safeApprove(address(treasury), claimAmount0Fee);
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol]: can put L209 and 210 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(token0()).safeApprove(address(treasury), type(uint256).max);
```

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+                treasury.receiveProfitFromVault(token0(), claimAmount0Fee);
+            }
+            if (_claimAmounts[1] > 0) {
+                uint256 claimAmount1Fee = _claimAmounts[1] * profitFeeBps / 10000;
+                _claimAmounts[1] -= claimAmount1Fee;
+                IERC20Upgradeable(token1()).safeApprove(address(treasury), 0);
+                IERC20Upgradeable(token1()).safeApprove(address(treasury), claimAmount1Fee);
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol]: can put L216 and 217 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(token0()).safeApprove(address(treasury), type(uint256).max);
```

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+
+    /// @notice Lend
+    /// @param _amount The amount of lend
+    function lend(uint256 _amount) external isOwner whenNotEmergency nonReentrant override {
+        IERC20Upgradeable(wantToken).safeTransferFrom(msg.sender, address(this), _amount);
+
+        if (manageFeeBps > 0 && address(treasury) != address(0)) {
+            uint256 manageFee = _amount * manageFeeBps / 10000;
+            IERC20Upgradeable(wantToken).safeApprove(address(treasury), 0);
+            IERC20Upgradeable(wantToken).safeApprove(address(treasury), manageFee);
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol]: can put L231 and 232 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(token0()).safeApprove(address(treasury), type(uint256).max);
```

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+                    amount1Min : 0,
+                    deadline : block.timestamp
+                    }));
+                    _balance0 = balanceOfToken(token0());
+                    _balance1 = balanceOfToken(token1());
+                }
+                if (_balance0 > 0 && _balance1 > 0) {
+                    //add liquidity
+                    nonfungiblePositionManager.increaseLiquidity(INonfungiblePositionManager.IncreaseLiquidityParams({
+                    tokenId : limitMintInfo.tokenId,
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol]: Can limit order add two tokens into LP V3 NFT ?

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+    }
+
+    /// @notice Redeem
+    /// @param _redeemShares The amount of shares to withdraw
+    /// @param _totalShares The total amount of shares owned by this strategy
+    /// @return _redeemBalance The balance of redeem
+    function redeem(uint256 _redeemShares, uint256 _totalShares) external isOwner override returns (uint256 _redeemBalance) {
+        _redeemBalance = redeemToVault(_redeemShares, _totalShares);
+        if (_redeemBalance > netMarketMakingAmount) {
+            netMarketMakingAmount = 0;
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol:286]: need add before `netMarketMakingAmount = 0;`
```
_redeemBalance = netMarketMakingAmount;
```

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+        address aCollateralToken = uniswapV3RiskOnHelper.getAToken(wantToken);
+        if (_redeemShares == _totalShares) {
+            harvest();
+            withdraw(baseMintInfo.tokenId, _redeemShares, _totalShares);
+            withdraw(limitMintInfo.tokenId, _redeemShares, _totalShares);
+            delete baseMintInfo;
+            delete limitMintInfo;
+            uint256 borrowTokenBalance = balanceOfToken(borrowToken);
+            if (currentBorrow > borrowTokenBalance) {
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol]: can put L318 and 319 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, type(uint256).max);
```

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 9]
```diff
@@ -0,0 +1,673 @@
...
+            withdraw(baseMintInfo.tokenId, _redeemShares, _totalShares);
+            withdraw(limitMintInfo.tokenId, _redeemShares, _totalShares);
+            //            uint256 redeemVaultWantTokenBalance = beforeWantTokenBalance * _redeemShares / _totalShares;
+            //            uint256 redeemVaultBorrowTokenBalance = beforeBorrowTokenBalance * _redeemShares / _totalShares;
+            //            uint256 redeemWantTokenBalance = afterWantTokenBalance - beforeWantTokenBalance;
+            uint256 redeemBorrowTokenBalance = balanceOfToken(borrowToken) - beforeBorrowTokenBalance;
+            uint256 redeemCurrentBorrow = currentBorrow * _redeemShares / _totalShares;
+            if (redeemCurrentBorrow > redeemBorrowTokenBalance) {
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
+                IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, balanceOfToken(wantToken));
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol]: can put L336 and 337 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(wantToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, type(uint256).max);
```

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 10]
```diff
@@ -0,0 +1,673 @@
...
+                IUniswapV3(RiskOnConstant.UNISWAP_V3_ROUTER).exactOutputSingle(IUniswapV3.ExactOutputSingleParams(wantToken, borrowToken, fee, address(this), block.timestamp, redeemCurrentBorrow - redeemBorrowTokenBalance, type(uint256).max, 0));
+            }
+            __repay(redeemCurrentBorrow);
+            __removeCollateral(aCollateralToken, IERC20(aCollateralToken).balanceOf(address(this)) * _redeemShares / _totalShares);
+        }
+        uint256 borrowTokenBalance = balanceOfToken(borrowToken);
+        if (borrowTokenBalance > 0) {
+            IERC20Upgradeable(borrowToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, 0);
+            IERC20Upgradeable(borrowToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, borrowTokenBalance);
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol]: can put L345 and 346 into function initialize,approve type(uint256).max amount for gas save
```
 IERC20Upgradeable(borrowToken).safeApprove(RiskOnConstant.UNISWAP_V3_ROUTER, type(uint256).max);
```

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 11]
```diff
@@ -0,0 +1,673 @@
...
+    /// @dev Shutdown the vault when an emergency occurs, cannot mint/burn.
+    /// Requirements: only vault manager can call
+    function setEmergencyShutdown(bool _active) external isVaultManager override {
+        emergencyShutdown = _active;
+        emit SetEmergencyShutdown(_active);
+    }
+
+    /// @dev Sets the manageFeeBps to the percentage of deposit that should be received in basis points.
+    function setManageFeeBps(uint256 _basis) external isVaultManager {
+        require(_basis <= 10, "basis cannot exceed 10%");
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol]: `_basis <= 10` should be `_basis <= 1000`

File: contracts/riskon/UniswapV3RiskOnVault.sol [hunk 12]
```diff
@@ -0,0 +1,673 @@
...
+import '@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/utils/math/SafeMath.sol";
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Factory.sol';
+import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol';
+import '@uniswap/v3-core/contracts/interfaces/IERC20Minimal.sol';
+import '@uniswap/v3-core/contracts/libraries/TickMath.sol';
+import '@uniswap/v3-core/contracts/libraries/SqrtPriceMath.sol';
+import "boc-contract-core/contracts/access-control/AccessControlMixin.sol";
```
Comment (baizetianxia) [inline contracts/riskon/UniswapV3RiskOnVault.sol:15]: `AccessControlMixin` contract need use v1.6.0-beta3

File: contracts/riskon/Treasury.sol [hunk 13]
```diff
@@ -0,0 +1,164 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+pragma solidity ^0.8.0;
+
+import '@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/token/ERC20/utils/SafeERC20Upgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol';
+import 'boc-contract-core/contracts/access-control/AccessControlMixin.sol';
```
Comment (baizetianxia) [inline contracts/riskon/Treasury.sol:9]: `AccessControlMixin` contract need use v1.6.0-beta3

File: contracts/riskon/VaultFactory.sol [hunk 14]
```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { Clones } from '@openzeppelin/contracts/proxy/Clones.sol';
+import 'boc-contract-core/contracts/access-control/AccessControlMixin.sol';
```
Comment (baizetianxia) [inline contracts/riskon/VaultFactory.sol:5]: `AccessControlMixin` contract need use v1.6.0-beta3

File: contracts/riskon/UniswapV3RiskOnVault.sol (hunk 6 above)
Comment (HHankCCui) [reply contracts/riskon/UniswapV3RiskOnVault.sol]: && to ||

**Review by baizetianxia — APPROVED** [review]

File: contracts/riskon/VaultFactory.sol [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+    /// @notice Gets one stablecoin investor by `_index`
+    /// @param _index The index of return investor on `stablecoinInvestorSet`
+    /// @return The investor address
+    function getStablecoinInvestorByIndex(uint256 _index) external view returns(address) {
+        return stablecoinInvestorSet.at(_index);
+    }
+
+    /// @notice Only for test.
+    ///         To be deleted before merge to master branch
+    function deleteVaultAddressMapForDebug(address _user, address _vaultImpl, uint256 _index) external {
```
Comment (baizetianxia) [inline contracts/riskon/VaultFactory.sol:156]: Only for test by frontend . To be deleted before merge

**Review by ZhangBrain — APPROVED** [review]

**Pull request merged**